daml-script/test/daml/upgrades/UpgradesTest.daml

```diff
@@ -0,0 +1,594 @@
+    , ("Call a V2 choice without package ID on a V1 contract over the ledger-api, expect V2 implementation used, and contract upgraded. (Nested)", inferredV2ChoiceV1ContractNested)
+    , ("(BROKEN) Call a V2 choice without package ID on a V1 contract over the ledger-api, with V2 unvetted, expect V1 implementation used, argument downgraded, daml-script upgrades return type. (Nested)", inferredV1ChoiceV1ContractWithoutV2Nested)
+    , ("Explicitly call a V1 choice on a V2 contract over the ledger-api, expect V1 implementation used, and contract downgraded. (Nested)", explicitV1ChoiceV2ContractNested)
+    , ("Explicitly call a V2 choice on a V2 contract over the ledger-api, expect V2 implementation used. (Nested)", explicitV2ChoiceV2ContractNested)
+    
+    , -- Invalid signatory/observer changes
+      ("Succeeds if the signatories don't change", unchangedSignatoryUpgrade)
+    , ("Fails if the signatories set gets larger", largerSignatoryUpgrade)
+    , ("Fails if the signatories set gets smaller", smallerSignatoryUpgrade)
+    , ("Succeeds if the observers don't change", unchangeObserverUpgrade)
+    , ("Fails if the observers set gets larger", largerObserverUpgrade)
+    , ("Fails if the observers set gets smaller", smallerObserverUpgrade)
+    , ("Succeeds if the observer set loses parties that are already signatories", canRemoveObserversThatAreSignatories)
+
+    , -- Invalid data upgrades (compile time checks repeated at runtime)
+      ("Fails if the template name changes", templateNameChanges)
+    , ("(BROKEN) Fails if a nested data-type's name changes", templateNameChangesNested)
+    , ("Fails if fields are removed", templateFieldsRemoved)
+    , ("Fails if non-optional fields are added", templateNonOptionalFieldsAdded)
+    , ("(BROKEN) Fails if fields are renamed", templateFieldsRenamed)
+    , ("Fails if nested fields are removed", templateFieldsRemovedNested)
+    , ("Fails if nested non-optional fields are added", templateNonOptionalFieldsAddedNested)
+    , ("(BROKEN) Fails if nested fields are renamed", templateFieldsRenamedNested)
+    , ("Succeeds if a nested variant is unchanged", templateVariantUnchanged)
+    , ("Fails if a nested variant is a removed case", templateVariantUpgradeToRemoved)
+    , ("Fails if a nested variant is an additional case when downgrading", templateVariantDowngradeFromNew)
+
+    , -- Edge behaviour
+      ("(BROKEN) Chooses the v1 contract if v2 is unvetted and package id is omitted.", packageSelectionChoosesUnvettedPackages)
+      -- TODO: Weird numeric case from Moritz/Remy - Requires disclosed contracts, daml3-script does not support this.
+      -- Cannot be tested.
+      -- I believe the case is - old contract with a Decimal, attempt to use it by disclosing and upgrading at the same time, something breaks
+    ]
+  , -- Multi participant tests
```

Comment:
Could consider breaking this out into a separate test as the category of error is more likely to indicate a canton issue more than an engine one.

Reply:
I'm somewhat using this to dogfood improving daml-script for testing, so I may implement something that handles this behaviour directly in script.
For now I'd like to keep the scala wrapper as simple as possible, knowing nothing of the tests inside